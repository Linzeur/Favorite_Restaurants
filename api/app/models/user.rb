class User < ApplicationRecord

  validates :name, presence: true
  validates :email, format: { with: /[a-zA-Z0-9]*@[a-zA-Z]*.[a-zA-Z0-9]*/, message: "Bad format"  },
                    uniqueness: true

  has_secure_password
  has_secure_token
  has_secure_token :reset_digest

  has_many :locations
  has_many :favorite_restaurants

  def self.valid_login?(email, password)
    user = find_by(email: email)
    user if user&.authenticate(password)
  end

  def invalidate_token
    update(token: nil)
  end
end
