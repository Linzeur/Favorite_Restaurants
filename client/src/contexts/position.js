import React from "react";

const PositionContext = React.createContext();

function PositionProvider(props) {
  const [position, setPosition] = React.useState([0, 0, false]);

  React.useEffect(() => {
    const watchId = navigator.geolocation.watchPosition(({ coords }) => {
      setPosition([coords.latitude, coords.longitude, true]);
      sessionStorage.setItem(
        "lastCurrentPosition",
        JSON.stringify({ lat: coords.latitude, lng: coords.longitude })
      );
    });
    return () => navigator.geolocation.clearWatch(watchId);
  }, [setPosition]);

  const value = {
    latitude: position[0],
    longitude: position[1],
    callAPI: position[2],
    onChange: (lat, lng, callAPI) => {
      setPosition([lat, lng, callAPI]);
    }
  };

  return <PositionContext.Provider value={value} {...props} />;
}

export { PositionProvider, PositionContext };
