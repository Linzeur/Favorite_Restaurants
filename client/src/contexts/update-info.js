import React from "react";

const UpdateListContext = React.createContext();

function UpdateListProvider(props) {
  const [updateList, setUpdateList] = React.useState({
    locations: 0,
    favorites: 0
  });

  const value = {
    locations: updateList.locations,
    favorites: updateList.favorites,
    onChange: key => {
      setUpdateList(prevState => {
        return { ...prevState, [key]: prevState[key] + 1 };
      });
    }
  };

  return <UpdateListContext.Provider value={value} {...props} />;
}

export { UpdateListProvider, UpdateListContext };
