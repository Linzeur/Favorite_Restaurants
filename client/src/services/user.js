import { apiUrl } from "../util";

const API_USERS = `${apiUrl}users`;

async function createUser(userData) {
  const response = await fetch(API_USERS, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(userData)
  });

  if (!response.ok) {
    const { errors } = await response.json();
    throw new Error(errors.message);
  }

  return response.json();
}

export { createUser };
