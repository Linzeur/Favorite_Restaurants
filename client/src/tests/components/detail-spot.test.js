import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";

import DetailSpot from "../../components/detail-spot";
import { UserProvider } from "../../contexts/user";
import { UpdateListProvider } from "../../contexts/update-info";

const spot = {
  place_id: "ChIJw42ff3DOBZERfN4MyMcqh5U",
  name: "Curayacu Querida Comidas Tipicas",
  vicinity: "Avenida Santa Rosa 1561, San MartÃ­n de Porres",
  lat: -11.9876734,
  lng: -77.09368529999999,
  formatted_phone_number: null,
  rating: 4.7,
  opening_hours: null,
  isFavorite: false
};

test("detail-spot component", async () => {
  fetch.mockResponseOnce(JSON.stringify(spot));
  const setModalFn = jest.fn();

  const { getByText, asFragment } = render(
    <UserProvider>
      <UpdateListProvider>
        <DetailSpot spotId={spot.place_id} showModalFn={setModalFn} />
      </UpdateListProvider>
    </UserProvider>
  );

  expect(asFragment()).toMatchSnapshot();

  const modalDetailSpot = getByText("Details");
  expect(modalDetailSpot).not.toBeNull();

  await wait(() => {});

  expect(asFragment()).toMatchSnapshot();
  const modalLabelName = getByText(spot.name);
  const modalLabelAddress = getByText(spot.vicinity);
  const modalLabelTelephone = getByText("Not Information");
  const modalLabelRating = getByText(spot.rating.toString());
  const modalLabelSchedule = getByText("Not Information about schedule");

  expect(modalLabelName).not.toBeNull();
  expect(modalLabelAddress).not.toBeNull();
  expect(modalLabelTelephone).not.toBeNull();
  expect(modalLabelRating).not.toBeNull();
  expect(modalLabelSchedule).not.toBeNull();
});

test("click in button close", () => {
  fetch.mockResponseOnce(JSON.stringify(spot));
  const setModalFn = jest.fn();

  const { getAllByRole, asFragment } = render(
    <UserProvider>
      <UpdateListProvider>
        <DetailSpot spotId={spot.place_id} showModalFn={setModalFn} />
      </UpdateListProvider>
    </UserProvider>
  );

  const buttonClose = getAllByRole("button")[0];

  fireEvent.click(buttonClose);

  expect(setModalFn).toHaveBeenCalledWith(null);
});
