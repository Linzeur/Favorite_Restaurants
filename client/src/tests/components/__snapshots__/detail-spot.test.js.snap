// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`detail-spot component 1`] = `
<DocumentFragment>
  <section
    class="css-w1nd5a-DetailSpot"
  >
    <form
      class="css-1ukvpoh-DetailSpot"
    >
      <div
        class="css-1p4pucu-DetailSpot"
      >
        <h2>
          Details
        </h2>
      </div>
      <hr />
      <div
        class="css-w40flo-Spinner"
      >
        <div
          class="css-lehx5n-Spinner"
        />
        <div
          class="css-wk0iwg-Spinner"
        />
        <div
          class="css-1xsj7ka-Spinner"
        />
      </div>
      <fieldset
        class="css-1xxup3v-DetailSpot"
      >
        <button
          class="css-u9qp7f-DetailSpot"
          type="button"
        >
          Close
        </button>
      </fieldset>
    </form>
  </section>
</DocumentFragment>
`;

exports[`detail-spot component 2`] = `
<DocumentFragment>
  <section
    class="css-w1nd5a-DetailSpot"
  >
    <form
      class="css-1ukvpoh-DetailSpot"
    >
      <div
        class="css-1p4pucu-DetailSpot"
      >
        <h2>
          Details
        </h2>
        <svg
          aria-label="Button with icon indicate spot is not favorite for you"
          class="css-wez6tj-DetailSpot"
          fill="currentColor"
          height="1em"
          role="button"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M458.4 64.3C400.6 15.7 311.3 23 256 79.3 200.7 23 111.4 15.6 53.6 64.3-21.6 127.6-10.6 230.8 43 285.5l175.4 178.7c10 10.2 23.4 15.9 37.6 15.9 14.3 0 27.6-5.6 37.6-15.8L469 285.6c53.5-54.7 64.7-157.9-10.6-221.3zm-23.6 187.5L259.4 430.5c-2.4 2.4-4.4 2.4-6.8 0L77.2 251.8c-36.5-37.2-43.9-107.6 7.3-150.7 38.9-32.7 98.9-27.8 136.5 10.5l35 35.7 35-35.7c37.8-38.5 97.8-43.2 136.5-10.6 51.1 43.1 43.5 113.9 7.3 150.8z"
          />
        </svg>
      </div>
      <hr />
      <fieldset
        class="css-1ihc67e"
      >
        <label
          class="css-19qo4af"
        >
          Name
        </label>
        <div
          class="css-1k3c23m"
        >
          <label>
            Curayacu Querida Comidas Tipicas
          </label>
        </div>
      </fieldset>
      <fieldset
        class="css-1ihc67e"
      >
        <label
          class="css-19qo4af"
        >
          Address
        </label>
        <div
          class="css-1k3c23m"
        >
          <label>
            Avenida Santa Rosa 1561, San Mart√≠n de Porres
          </label>
        </div>
      </fieldset>
      <fieldset
        class="css-1ihc67e"
      >
        <label
          class="css-19qo4af"
        >
          Telephone
        </label>
        <div
          class="css-1k3c23m"
        >
          <label>
            Not Information
          </label>
        </div>
      </fieldset>
      <fieldset
        class="css-1ihc67e"
      >
        <label
          class="css-19qo4af"
        >
          Rating
        </label>
        <div
          class="css-1k3c23m"
        >
          <label>
            4.7
          </label>
        </div>
      </fieldset>
      <fieldset
        class="css-1xxup3v-DetailSpot"
      >
        <label
          class="css-6ius79-DetailSpot"
        >
          Schedule
        </label>
        <div
          class="css-nzcmyo-DetailSpot"
        >
          <label>
            Not Information about schedule
          </label>
        </div>
      </fieldset>
      <fieldset
        class="css-1xxup3v-DetailSpot"
      >
        <button
          class="css-u9qp7f-DetailSpot"
          type="button"
        >
          Close
        </button>
      </fieldset>
    </form>
  </section>
</DocumentFragment>
`;
